StRoot/StFcsDbMaker/StFcsDb.cxx

```diff
@@ -1979,3 +1979,107 @@ void StFcsDb::readGainCorrFromText(){
+                     id,g2ttrk[i].id,g2ttrk[i].eg_pid,g2ttrk[i].ge_pid,g2ttrk[i].start_vertex_p,
+		     g2ttrk[i].next_parent_p,g2ttrk[i].e)<<endm;
+    i = g2ttrk[i].next_parent_p - 1;
+  }
+  if(mDebug>3) LOG_INFO<<Form("  BackTrace from=%3d id=%3d Epid=%4d Gpid=%3d Vtx=%3d Parent=%3d E=%6.2f Primary!!",
+			      id,g2ttrk[i].id,g2ttrk[i].eg_pid,g2ttrk[i].ge_pid,g2ttrk[i].start_vertex_p,
+			      g2ttrk[i].next_parent_p,g2ttrk[i].e)<<endm;
+  return i + 1;
+}
+    
+const g2t_track_st* StFcsDb::getParentG2tTrack(StFcsHit* h, g2t_track_st* g2ttrk, float& fraction, int& ntrk, unsigned int order){
+  const vector<pair<unsigned int, float>>& gt = h->getGeantTracks();  
+  ntrk=gt.size();
+  float detot=0;
+  if(order > gt.size()) {fraction=0; return 0;}
+  for(int itrk=0; itrk<ntrk; itrk++) detot += gt[itrk].second;
+  fraction = gt[order].second / detot;
+  if(mDebug>3){
+    for(unsigned int jtrk=0; jtrk<ntrk; jtrk++){
+      LOG_INFO << Form("Hit's parent G2T Track %3d id=%3d dE=%f",jtrk,gt[jtrk].first,gt[jtrk].second)<<endm;
+    }
+  }
+  return &g2ttrk[gt[order].first-1];
+}
+
+const g2t_track_st* StFcsDb::getPrimaryG2tTrack(StFcsHit* h, g2t_track_st* g2ttrk, float& fraction, int& ntrk,unsigned int order){
+  const vector<pair<unsigned int, float>>& gt = h->getGeantTracks();
+  vector<pair<unsigned int, float>> primary;
+  int n=gt.size();
+  float detot=0;
+  for(int itrk=0; itrk<n; itrk++){
+    float de=gt[itrk].second;
+    unsigned int id=backTraceG2tTrack(gt[itrk].first, g2ttrk);
+    int found=0;
+    unsigned int np=primary.size();
+    for(unsigned int jtrk=0; jtrk<np; jtrk++){
+      if(primary[jtrk].first == id) {primary[jtrk].second += de; found=1; break;}
+    }
+    if(found==0) primary.push_back(make_pair(id,de));
+    detot+=de;
+  }
+  ntrk = primary.size();
+  if(order > ntrk) {fraction=0; return 0;}
+  std::nth_element(primary.begin(), primary.begin() + order, primary.end(), 
+	    [](const pair<unsigned int,float>&a, const pair<unsigned int,float>&b){
+	      return b.second < a.second;
+	    });
+  fraction = primary[order].second / detot;
+  return &g2ttrk[primary[order].first-1];
+}
+
+const g2t_track_st* StFcsDb::getParentG2tTrack(StFcsCluster* c, g2t_track_st* g2ttrk, float& fraction, int& ntrk, unsigned int order){
+  getG2tTrack(c,g2ttrk,fraction,ntrk,order,0);
+}
+
+const g2t_track_st* StFcsDb::getPrimaryG2tTrack(StFcsCluster* c, g2t_track_st* g2ttrk, float& fraction, int& ntrk, unsigned int order){
+  getG2tTrack(c,g2ttrk,fraction,ntrk,order,1);
+}
+
+const g2t_track_st* StFcsDb::getG2tTrack(StFcsCluster* c, g2t_track_st* g2ttrk, float& fraction, int& ntrk, unsigned int order, int mode){
+  float detot=0;
+  vector<pair<unsigned int, float>> parents;
+  StPtrVecFcsHit& hits = c->hits();  
+  int nhit = hits.size();
+  for(int ihit=0; ihit<nhit; ihit++){    
+    const vector<pair<unsigned int, float>>& gt = hits[ihit]->getGeantTracks();
```

Comment:
Is there a reason why you prefer to use indices over range for loops?

```c++
  for (const StFcsHit* hit : c->hits()) {
    const auto& gt = hit->getGeantTracks();
```

Looks less verbose

Reply:
I prefer indices over iterator & range since I often end up wanting to have random access & indexes later in the loop. But this case it may be ok with range. Change coming.